QuantConnect.ExanteBrokerage/ExanteBrokerage.cs

```diff
@@ -0,0 +1,839 @@
+            var asks = exanteTickShort.Ask.ToList();
+            return new Tick(time, symbol, "", "",
+                bids.IsNullOrEmpty() ? decimal.Zero : bids[0].Size,
+                bids.IsNullOrEmpty() ? decimal.Zero : bids[0].Price,
+                asks.IsNullOrEmpty() ? decimal.Zero : asks[0].Size,
+                asks.IsNullOrEmpty() ? decimal.Zero : asks[0].Price);
+        }
+
+        #endregion
+
+        private void OnUserMessage(ExanteOrder exanteOrder)
+        {
+            Order order;
+            if (!_orderMap.TryGetValue(exanteOrder.OrderId, out order))
+            {
+                return;
+            }
+
+            var exanteSymbol = Client.GetSymbol(exanteOrder.OrderParameters.SymbolId);
+            var symbol = ConvertSymbol(exanteSymbol);
+
+            var security = _securityProvider.GetSecurity(symbol);
+            var orderFeeParameters =  new OrderFeeParameters(security, order);
+            var fee = _feeModel.GetOrderFee(orderFeeParameters);
+
+            var orderEvent = new OrderEvent(order, DateTime.UtcNow, fee)
+            {
+                Status = ConvertOrderStatus(exanteOrder.OrderState.Status),
+            };
+            if (exanteOrder.OrderState.Status == ExanteOrderStatus.Filled)
+            {
+                orderEvent.FillQuantity = exanteOrder.OrderParameters.Quantity;
+                _messageHandler.WithLockedStream(() => { _orderMap.TryRemove(exanteOrder.OrderId, out _); });
+            }
+
+            OnOrderEvent(orderEvent);
+        }
+
+        private IEnumerable<BaseData> GetCandlesHistory(Data.HistoryRequest request)
+        {
+            var symbol = SymbolMapper.GetBrokerageSymbol(request.Symbol);
+
+            var exanteTickType = request.TickType switch
+            {
+                TickType.Quote => ExanteTickType.Quotes,
+                TickType.Trade => ExanteTickType.Trades,
+                TickType.OpenInterest => throw new ArgumentException(),
+                _ => throw new ArgumentOutOfRangeException()
+            };
+
+            var exanteTimeframe = request.Resolution switch
+            {
+                Resolution.Tick => throw new ArgumentException(),
+                Resolution.Second => throw new ArgumentException(),
+                Resolution.Minute => ExanteCandleTimeframe.Minute1,
+                Resolution.Hour => ExanteCandleTimeframe.Hour1,
+                Resolution.Daily => ExanteCandleTimeframe.Day1,
+                _ => throw new ArgumentOutOfRangeException()
+            };
+
+            var history = Client.GetCandles(
+                symbol,
+                exanteTimeframe,
+                from: request.StartTimeUtc,
+                to: request.EndTimeUtc,
+                tickType: exanteTickType
+            ).Data.ToList();
+
+            var period = request.Resolution.ToTimeSpan();
+
+            foreach (var kline in history)
+            {
+                yield return new TradeBar()
```

Comment:
> DataType = MarketDataType.TradeBar,

Should use the QuoteBar and TradeBar constructors instead of the public setters, to avoid these issues.
Does exante provide bid/ask bar information for the quoteBar?

Reply:
Attempted to fix it at 372cb7dda86f2f05c70ce792139ddc2b1faa1871. I didn't manage to create `QuoteBar` from constructor because I'm not sure where to get bid/ask. I expect you asked about bid/ask of QuoteBar for that reason. `ExanteCandle` doesn't have bid/ask

```c#
public class ExanteCandle
{
  /// <summary>Candle close price</summary>
  public Decimal Close { get; set; }
  /// <summary>Candle open price</summary>
  public Decimal Open { get; set; }
  /// <summary>Candle low price</summary>
  public Decimal Low { get; set; }
  /// <summary>Candle high price</summary>
  public Decimal High { get; set; }
  /// <summary>
  /// Total volume for specified period. Appears and required only for trade candle request
  /// </summary>
  public Decimal? Volume { get; set; }
  /// <summary>Candle date</summary>
  [JsonProperty(PropertyName = "timestamp")]
  [JsonConverter(typeof (TimestampConverter))]
  public DateTime Date { get; set; }
}
```

There are methods [getQuoteLast](https://api-live.exante.eu/api-docs/#operation/getQuoteLast) and [getQuoteStream](https://api-live.exante.eu/api-docs/#operation/getQuoteStream) that return bid/ask. But that means extra REST-API call that is not what you'd like.

Reply:
> ExanteCandle

This seems like the Lean `TradeBar` not `QuoteBar`. 

> There are methods [getQuoteLast](https://api-live.exante.eu/api-docs/#operation/getQuoteLast) and [getQuoteStream](https://api-live.exante.eu/api-docs/#operation/getQuoteStream) that return bid/ask. But that means extra REST-API call that is not what you'd like.

The problem isn't the api call if required, but those endpoints don't return historical data? 
If exante doesn't return quote information, that's okay it should `yield return` detecting that case, see similar handling:
https://github.com/QuantConnect/Lean/blob/master/Brokerages/GDAX/GDAXBrokerage.cs#L339
 https://github.com/QuantConnect/Lean/blob/master/Brokerages/Bitfinex/BitfinexBrokerage.cs#L338